sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/HadoopFileLinesReader.scala

```diff
@@ -32,7 +32,9 @@ import org.apache.hadoop.mapreduce.task.TaskAttemptContextImpl
  * in that file.
  */
 class HadoopFileLinesReader(
-    file: PartitionedFile, conf: Configuration) extends Iterator[Text] with Closeable {
+    file: PartitionedFile,
+    lineSeparator: Option[String],
```

Comment:
This should not be optional.

Reply:
I saw a discussion in the Impala. https://issues.apache.org/jira/browse/IMPALA-1578
 
> When the line delimiter is '\n', any of the follow sequences will count as a delimiter: "\n", "\r\n", or "\r". Accepting a single "\r" is pretty strange, but that's what Hive does so we emulate this behavior.

Could you check whether Hive is behaving like this?

Reply:
I am willing to do so but I am curious of the purpose of checking so. Do you want some test cases about this concern?

Reply:
Is Hive using Hadoop's `LineRecordReader`? How does Hive support it?

If possible, we always try to behave the same like Hive.

Reply:
Thanks for clarifying it. Here is my investigation:

> When the line delimiter is '\n', any of the follow sequences will count as a delimiter: "\n", "\r\n", or "\r"

With this input:

```
a\nb\r\nc\rd
```

Case with `\n`:

```sql
CREATE EXTERNAL TABlE tbl(value STRING)
ROW FORMAT DELIMITED LINES TERMINATED BY '\n'
STORED AS TEXTFILE LOCATION '...';
```

```sql
SELECT value FROM tbl;
```

produced

```
a
b
c
d
```

This looks incorrect. I _guess_ `\n` is not being set and it looks working as the default behaviour in `LineRecordReader`.


> Accepting a single "\r" is pretty strange, but that's what Hive does so we emulate this behavior.

Case with `\r`:

```sql
CREATE EXTERNAL TABlE tbl(value STRING) 
ROW FORMAT DELIMITED LINES TERMINATED BY '\r'
STORED AS TEXTFILE LOCATION '...';
```

produced

```
FAILED: SemanticException 2:41 LINES TERMINATED BY only supports newline '\n' right now. Error encountered near token ''\r''
...
org.apache.hadoop.hive.ql.parse.SemanticException: 2:41 LINES TERMINATED BY only supports newline '\n' right now. Error encountered near token ''\r''
```

This looks related with https://issues.apache.org/jira/browse/HIVE-5999

and these lines:

https://github.com/apache/hive/blob/696be9f52dfc6fb59c24de19726b4460100fc9ba/ql/src/java/org/apache/hadoop/hive/ql/parse/BaseSemanticAnalyzer.java#L198-L203

if I am not mistaken.

I am curious how this case was tested in the JIRA. If this test used `textinputformat.record.delimiter`, then, this seems Hadoop's property, which is basically the same thing as what I am doing here.


> Is Hive using Hadoop's LineRecordReader?

In the case above, the input format was `org.apache.hadoop.mapred.TextInputFormat`, which uses `LineRecordReader`. 

> How does Hive support it?

It looks Hive tries to support it by `LINES TERMINATED BY '\r'` https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL#LanguageManualDDL-CreateTableCreate/Drop/TruncateTable. I could not find other (formal) ways.

Reply:
**\n = LF:**    
> Multics, Unix and Unix-like systems (Linux, macOS, FreeBSD, AIX, Xenix, etc.), BeOS, Amiga, RISC OS, and others[1]

**\r\n CR+LF:** 
> Microsoft Windows, DOS (MS-DOS, PC DOS, etc.), DEC TOPS-10, RT-11, CP/M, MP/M, Atari TOS, OS/2, Symbian OS, Palm OS, Amstrad CPC, and most other early non-Unix and non-IBM operating systems

**\r = CR:**    
> Commodore 8-bit machines, Acorn BBC, ZX Spectrum, TRS-80, Apple II family, Oberon, the classic Mac OS, MIT Lisp Machine and OS-9

Reply:
It sounds like Hive's behavior is reasonable. For most external users, `\n` means a new line. Thus, it should match any of these three options: `LF`, `CR+LF` and `CR`. WDYT?

Reply:
I was thinking `\n` means the first case of your comment above as it is set by the user explicitly. So, I thought If it is not given, it covers three cases of newlines by default. If we use `\n` to deal with three cases above, wouldn't we are unable to cover, the arguably corner case of only handling `\n`?

Reply:
So far, following Hive is the safest. If users complain about it, we can behave differently from Hive with a new SQLConf.

Reply:
OK. Will change this but I should say this way looks incorrect to me and this behaviour should be discussed and possibly updated in the near future.